Metagenomics/Remove_human_reads_from_raw_data/Workflow_Documentation/NF_MGRemoveHumanReads-A/workflow_code/Remove_Human_Reads.nf

```diff
@@ -0,0 +1,161 @@
+
+    tar -xzvf ${params.human_db_name}.tar.gz
+    rm ${params.human_db_name}.tar.gz
+
+    """
+
+}
+
+// Process for paired-end (PE) read analysis with Kraken2
+process PE_kraken2 {
+
+    container params.kraken2container
+    tag "$sample_id"
+    publishDir "$params.kraken_output_dir", pattern: "*.{txt,tsv}"
+    publishDir "$params.kraken_output_dir/reads", pattern: "*.fastq.gz"
+
+    input:
+    path database 
+    tuple val(sample_id), path(reads_ch)
+    
+
+    output:
+    path "${sample_id}-kraken2-output.txt"
+    path "${sample_id}-kraken2-report.tsv"
+    path "${sample_id}*.gz"
+
+    script:
+    """
+    kraken2 --db $database --gzip-compressed \
+            --threads 2 --use-names --paired  \
+            --output ${sample_id}-kraken2-output.txt \
+            --report ${sample_id}-kraken2-report.tsv \
+            --unclassified-out "${sample_id}${params.PE_reads_out_suffix}" \
+            ${reads_ch[0]} ${reads_ch[1]}
+
+    gzip ${sample_id}*.fastq
+    """
+}
+
+// Process for single-end (SE) read analysis with Kraken2
+process SE_kraken2 {
+    
+    container params.kraken2container 
+    tag "$sample_id"
+    publishDir "$params.kraken_output_dir", pattern: "*.{txt,tsv}"
+    publishDir "$params.kraken_output_dir/reads", pattern: "*.fastq.gz"
+
+    input:
+    path database 
+    tuple val(sample_id), path(reads_ch)
+
+    output:
+    path "${sample_id}-kraken2-output.txt"
+    path "${sample_id}-kraken2-report.tsv"
+    path "${sample_id}${params.SE_reads_out_suffix}.gz"
+
+    script:
+    """
+    kraken2 --db $database --gzip-compressed --threads 2 --use-names  \
+            --output ${sample_id}-kraken2-output.txt \
+            --report ${sample_id}-kraken2-report.tsv \
+            --unclassified-out "${sample_id}${params.SE_reads_out_suffix}" \
+            ${reads_ch[0]}
+
+    gzip ${sample_id}${params.SE_reads_out_suffix}
+    """
+}
+
+
+// Main workflow logic
+workflow {
+
+// Conditionally download the human reads database if requested
+if(params.DL_kraken == true){
+    log.info "\nPreparing to download new human reads database"
+    database_ch =  set_up_human_db()
+    database_ch.view{"database path: ${it}"}
+}
+
+else {
+    log.info "\nAccessing previous human reads database"
+    database_ch = Channel.value(params.human_db_path)
+    database_ch.view{"database path: ${it}"}
+}
+
+// Process reads based on whether they are single-end or paired-end
+if(params.single_end == true) {
+    log.info "\nReading Single-end data from ${params.reads_dir}\n"
+    
+    // Specified reads handling (mimics Channel.fromFilePairs() method's output, but with SE)
+    if (params.specify_reads) {
+        reads_ch = Channel
+        .fromPath("${params.sample_id_list}")
+        .splitText()
+        .map { it.trim() }
+        .map { sample_id ->
+            def files = file("${params.reads_dir}${sample_id}${params.SE_reads_suffix}")

```

Comment:
Don't assume that users will put the path separator on the end of the path they specify.
```suggestion
            def files = file("${params.reads_dir}/${sample_id}${params.SE_reads_suffix}")

```